File: NgheAn_Vinh_2324/Cau4/CHONPHONG.py
def main():
    # Đọc dữ liệu từ file CHONPHONG.INP
    with open("CHONPHONG.INP", "r") as infile:
        N, K = map(int, infile.readline().split())
        S = infile.readline().strip()

    # Tìm vị trí các phòng trống
    empty_rooms = [i for i in range(N) if S[i] == '0']

    min_distance = N  # Khởi tạo khoảng cách nhỏ nhất có thể

    # Duyệt qua các nhóm phòng trống với kích thước K + 1
    for i in range(len(empty_rooms) - K):
        teacher_room = empty_rooms[i + K // 2]
        max_distance = max(teacher_room - empty_rooms[i], empty_rooms[i + K] - teacher_room)
        min_distance = min(min_distance, max_distance)

    # Ghi kết quả vào file CHONPHONG.OUT
    with open("CHONPHONG.OUT", "w") as outfile:
        outfile.write(f"{min_distance}\n")

if __name__ == "__main__":
    main()

File: THCS/2023_2024/NgheAn_Vinh_2324/Cau4/CHONPHONG.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <deque>
#include <algorithm>

using namespace std;

int main() {
    // Đọc dữ liệu từ file CHONPHONG.INP
    ifstream infile("CHONPHONG.INP");
    int N, K;
    infile >> N >> K;

    string S;
    infile >> S;
    infile.close();

    // Tạo danh sách vị trí các phòng trống
    vector<int> empty_rooms;
    for (int i = 0; i < N; ++i) {
        if (S[i] == '0') {
            empty_rooms.push_back(i);
        }
    }

    int min_distance = N; // Khởi tạo khoảng cách lớn nhất có thể
    int num_empty = empty_rooms.size();

    // Duyệt qua các nhóm phòng trống với kích thước K+1
    for (int i = 0; i <= num_empty - K - 1; ++i) {
        // Phòng của Thầy Nam là phòng ở giữa nhóm
        int teacher_room = empty_rooms[i + K / 2];
        // Khoảng cách lớn nhất từ phòng của Thầy Nam đến phòng của học sinh xa nhất
        int max_distance = max(teacher_room - empty_rooms[i], empty_rooms[i + K] - teacher_room);
        min_distance = min(min_distance, max_distance);
    }

    // Ghi kết quả ra file CHONPHONG.OUT
    ofstream outfile("CHONPHONG.OUT");
    outfile << min_distance << endl;
    outfile.close();

    return 0;
}
